prereq: make existing binary check work for sdk as well

KanjiMonster opened the pull request:
To avoid replacing host built binaries with symlinks again, a check for an appropriate stamp was added in 729909c07f ("prereq-build: do not replace binaries with symlinks"). Unfortunately the stamp directory does not exist in the packages SDK, so the fix was ineffective there.

This caused the packages builders to e.g. use the host tar again, which in turn made the tarballs created different since it may lack reproducibility fixes, or implement these differently, causing spurious hash failures on source repository based packages.

Fix this by dropping the stamp dir check, and just check that the file is usable.

Now that most cases do the same thing in SetupHostCommand, we can merge them
together into one. To allow moving the generic symlink check, invert the
check and let it check for relative links by matching on link targets
that do not start with a slash.

This then allows us to also drop the absolute link case, shortening the
case statement further.

This reorders the check to

* if it is not a symlink, do not change it
* if it is a symlink and it points to the found command, do not change it
* if it is a symlink with a relative path, do not change it
* else, update/replace it

Fixes: 729909c07f ("prereq-build: do not replace binaries with symlinks")

KanjiMonster committed:
prereq: make existing binary check work for sdk as well

To avoid replacing host built binaries with symlinks again, a check for
an appropriate stamp was added in 729909c07f ("prereq-build: do not
replace binaries with symlinks"). Unfortunately the stamp directory does
not exist in the SDK, so the fix was ineffective there.

This caused the packages builders to e.g. use the host tar again, which
in turn made the tarballs created different since it may lack
reproducibility fixes, or implement these differently, causing spurious
hash failures on source repository based packages.

Fix this by dropping the stamp dir check, and just check that the file
is usable.

Fixes: 729909c07f ("prereq-build: do not replace binaries with symlinks")
Signed-off-by: Jonas Gorski <jonas.gorski@gmail.com>

computersforpeace commented:
> This caused the packages builders to e.g. use the host tar again, which in turn made the tarballs created different since it lacks some reproducibility fixes, causing spurious hash failures.

For the record, the host tar does *not* lack reproducibility -- it just reproduces differently ;)

Whereas openwrt just hacks in "root:root" user/group in for LongLinks, standard tar simply omits them when used with --numeric-owner.

But this is just a nitpick in wording. It's effectively similar, whether it's reproducing a different tarball, or not reproducing at all.

(I still think we should just drop the openwrt patch, because it's a hack, and it applies its methods inconsistently -- only to LongLink, and not regular (<100 char name) links. But that's a different question.)

ynezz commented:
@mpratt14  FYI

KanjiMonster commented:
> > This caused the packages builders to e.g. use the host tar again, which in turn made the tarballs created different since it lacks some reproducibility fixes, causing spurious hash failures.
> 
> For the record, the host tar does _not_ lack reproducibility -- it just reproduces differently ;)
> 
> Whereas openwrt just hacks in "root:root" user/group in for LongLinks, standard tar simply omits them when used with --numeric-owner.
> 
> But this is just a nitpick in wording. It's effectively similar, whether it's reproducing a different tarball, or not reproducing at all.

The advantage of PRs, I can just edit the commit message. There, reworded it a bit. Now better? ;P

> (I still think we should just drop the openwrt patch, because it's a hack, and it applies its methods inconsistently -- only to LongLink, and not regular (<100 char name) links. But that's a different question.)

While I understand the sentiment, this would likely require recalculating and potentially updating all hashes for packages using source repositories as checkouts, and uploading new source packages.

And we cannot do that for main alone, as all branches potentially using the same source packages would need to be updated as well so *their* downloads don't fail. And then update all their source packages etc, which might trigger the need to update even earlier versions etc.

And it likely would break building any tagged release, since their package hashes would now be wrong for the updated tarballs.

We could maybe avoid the backporting part by using a different name for upstream tar generated tarballs, then the ones with the new hash and the old hash can co-exist.

Not sure if it is worth the headache though.

computersforpeace commented:
Ack. Looks better

mpratt14 commented:
I apologize for overlooking how my changes affect the SDK

can we add a wildcard statement (looking for the stamp directory) for the first part of the case statement? If I had realized this problem ahead of time I probably would have done that...

If that doesn't sound good, at the very least the check should remain in the case statement and not be moved to the top, otherwise it would also undo the reason these changes were done in the first place

see
f75204036ccc56700df18258602cc65726dd653b

mpratt14 commented:
in other words

`$(if $(wildcard <stampdir>), ...) /`

KanjiMonster commented:
> in other words
> 
> `$(if $(wildcard <stampdir>), ...) /`

Updating symlinks still works, it just won't update regular files to symlinks anymore. `-f` only returns true if the file exists and is a regular file, and symlink aren't regular files. So for any symlinks (or not existing files) it will still continue on to the case statement.

Or am I missing anything?

computersforpeace commented:
> `-f` only returns true if the file exists and is a regular file, and symlink aren't regular files.

Erm:

```
$ touch a; ln -s a b; [ -f b ] && echo "symlink is a file"
symlink is a file
```

You might be looking for `-L` combined with `-f`:

```
$ touch a; ln -s a b
if [ -f b ] && ! [ -L b ]; then
  echo "b is a regular file"
else
  echo "b is not a regular file"
fi
b is not a regular file
```

I get that the docs are confusing; `bash` has a better description of "regular file":

https://tldp.org/LDP/abs/html/fto.html

"file is a regular file (not a directory or device file)"

mpratt14 commented:
for those reasons and more, I did change the file check to `-x` (executable or not) since it is forced to follow the link

> Or am I missing something?

the prereq checks won't be run at all if the file check is placed at the top AND passes (exit 0) but only in the cases where it would otherwise fail first (exit 1) if the current link doesn't match what the prereq check decides it should be, and that would allow an old version of, for example, python.

all we need is a single line change so that stampfiles are only checked when the stamp directory exists, assuming that adding the stamps to the SDK is not an option...

(I kind of prefer the latter but don't really care)

KanjiMonster commented:
> > `-f` only returns true if the file exists and is a regular file, and symlink aren't regular files.
> 
> Erm:
> 
> ```
> $ touch a; ln -s a b; [ -f b ] && echo "symlink is a file"
> symlink is a file
> ```

Yeah, must have botched my testing (I usually just test my assumption instead of reading docs). Probably only tested with broken links.

> all we need is a single line change so that stampfiles are only checked when the stamp directory exists, assuming that adding the stamps to the SDK is not an option...

Can we just instead drop the stamp check? I'd rather have code that behaves the same in !SDK and SDK. I think the likelyhood of ever needing to replace a self built binary with a host binary is close to zero.

This would also allow simplifying the code, since then all cases except `"-> /"*` are the same.

mpratt14 commented:
I was hoping to keep the stampfile check to make it as robust as possible, but if you think it makes no difference then go ahead and get rid of the line.

I suppose it only handles an odd case where the user manually uninstalls a single host tools program and then changes branch, but the user should be able to know what's going on.

mpratt14 commented:
but leave the cases alone, it's done this way (if I remember correctly...) to handle some weird situations with macOS having a non-gnu install, so that a symlink called install pointing to ginstall is needed

KanjiMonster committed:
prereq: merge common cases in SetupHostCommand

Now that most cases do the same thing in SetupHostCommand, merge them
together into one. To allow moving the generic symlink check, invert the
check and let it check for relative links by matching on link targets
that do not start with a slash.

This then allows us to also drop the absolute link case, shortening the
case statement further.

This reorders the check to

* if it is not a symlink, do not change it
* if it is a symlink and it points to the found command, do not change it
* if it is a symlink with a relative path, do not change it
* else, update/replace it

Signed-off-by: Jonas Gorski <jonas.gorski@gmail.com>

KanjiMonster commented:
Updated the check to not do a stamp check, and merged the switch cases that are doing the same thing (in a separate commit, as only the fix should be backported).

mpratt14 commented:
looks good to me now

nice use of character class I never would have thought of it
